src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp

```diff
@@ -234,31 +233,86 @@ size_t ShenandoahCardCluster::get_last_start(size_t card_index) const {
   return _object_starts[card_index].offsets.last;
 }
 
-// Given a card_index, return the starting address of the first block in the heap
-// that straddles into this card. If this card is co-initial with an object, then
-// this would return the first address of the range that this card covers, which is
-// where the card's first object also begins.
-HeapWord* ShenandoahCardCluster::block_start(const size_t card_index) const {
+HeapWord* ShenandoahCardCluster::first_object_start(const size_t card_index, const ShenandoahMarkingContext* const ctx,
+                                                    HeapWord* tams, HeapWord* end_range_of_interest) const {
 
   HeapWord* left = _rs->addr_for_card_index(card_index);
+  assert(left < end_range_of_interest, "No meaningful work to do");
+  ShenandoahHeapRegion* region = ShenandoahHeap::heap()->heap_region_containing(left);
 
 #ifdef ASSERT
   assert(ShenandoahHeap::heap()->mode()->is_generational(), "Do not use in non-generational mode");
-  ShenandoahHeapRegion* region = ShenandoahHeap::heap()->heap_region_containing(left);
   assert(region->is_old(), "Do not use for young regions");
   // For HumongousRegion:s it's more efficient to jump directly to the
   // start region.
   assert(!region->is_humongous(), "Use region->humongous_start_region() instead");
 #endif
+
+  HeapWord* right = MIN2(region->top(), end_range_of_interest);
```

Comment:
This is a safe thing to do, but doesn't the caller already establish the invariant that
`region->top() >= end_range_of_interest` ? Can we just assert that? (And rename the formal parameter name from `end_range_of_interest` to `right`?)

If so, we might also want to change the name of the formal parameter from `card_index` to `left`, and change it to be a (card-aligned) heap address for symmetry in the API.